### PR TITLE
Coord-based access to local NArray elements

### PR DESCRIPTION
**Pull request by fuchsto**
Addresses #168

### CODE REVIEW

File: dash/include/dash/algorithm/Generate.h [hunk 1]
```diff
@@ -46,6 +46,8 @@ void generate (
   auto llast  = lrange.end;
 
   std::generate(lfirst, llast, gen);
+
+  first.team().barrier();
```
Comment (fmoessbauer) [inline dash/include/dash/algorithm/Generate.h]: Are these barriers really intended? This accounts for all dash algorithms.

File: dash/include/dash/matrix/internal/LocalMatrixRef-inl.h [hunk 2]
```diff
@@ -321,14 +323,21 @@ ::at(
...
+  DASH_LOG_TRACE("LocalMatrixRef.at()",
+                 "ndim:",     NumDim,
+                 "curdim:",   CUR,
+                 "l_coords:",   _refview._coord,
+                 "l_viewspec:", _refview._l_viewspec);
   return local_at(
            _refview._mat->_pattern.local_at(
              _refview._coord,
-             _refview._viewspec));
+             _refview._l_viewspec));
```
Comment (fmoessbauer) [inline dash/include/dash/matrix/internal/LocalMatrixRef-inl.h:340]: This is the actual fix, isn't it?

**Review by fmoessbauer — APPROVED** [review]
Good Work!

File: dash/include/dash/matrix/internal/LocalMatrixRef-inl.h (hunk 2 above)
Comment (fuchsto) [reply dash/include/dash/matrix/internal/LocalMatrixRef-inl.h:340]: Supposedly this change is all that is required to fix the defect, yes.
`_coord` refers to local index range, `_viewspec` to global index range (global view to local index range, that is). Should be viewspec in local index range, of course. Fixed tests on my machines.

Also added unit test to cover this use case.

File: dash/include/dash/algorithm/Generate.h (hunk 1 above)
Comment (fuchsto) [reply dash/include/dash/algorithm/Generate.h]: There was a discussion on this some time ago. Didn't we decide for barriers?

Comment (fmoessbauer) [reply dash/include/dash/algorithm/Generate.h]: Ok, I have no idea ... We should document that in the algorithm section. A quick survey showed that some other algorithms (e.g. [fill](https://github.com/dash-project/dash/blob/f716024fe3d8a5408dbaa2551ba1c8e0153fab42/dash/include/dash/algorithm/Fill.h#L35)) do not use barriers.

Comment (fuchsto) [reply dash/include/dash/algorithm/Generate.h]: Ok, will remove the barrier before merging.

**Pull request merged**